KIPs/kip-37.md

```diff
@@ -0,0 +1,908 @@
+- An example:
+  1. An approved user calls a function such as `function myTransferFrom(address _from, address _to, uint256[] calldata _ids, uint256[] calldata _values);`.
+  2. `myTransferFrom` updates the balances for `_from` and `_to` addresses for all `_ids` and `_values`.
+  3. `myTransferFrom` emits `TransferBatch` with the details of what was transferred from address `_from` to address `_to`.
+  4. `myTransferFrom` checks if `_to` is a contract address and determines that it is so (if not, then the transfer can be considered successful).
+  5. `myTransferFrom` calls `onKIP37BatchReceived` on `_to` and it reverts or returns an unknown value (if it had returned `bytes4(keccak256("onKIP37BatchReceived(address,address,uint256[],uint256[],bytes)"))` the transfer can be considered successful).
+  6. At this point `myTransferFrom` SHOULD revert the transaction immediately as receipt of the token(s) was not explicitly accepted by the `onKIP37BatchReceived` function.
+  7. If however `myTransferFrom` wishes to continue it MUST call `supportsInterface(0x7cc2d017)` on `_to` and if it returns the constant value `true` the transaction MUST be reverted, as it is now known to be a valid receiver and the previous acceptance step failed.
+     - NOTE: You could have called `supportsInterface(0x7cc2d017)` at a previous step if you wanted to gather and act upon that information earlier, such as in a hybrid standards scenario.
+  8. If the above call to `supportsInterface(0x7cc2d017)` on `_to` reverts or returns a value other than the constant value `true` the `myTransferFrom` function MAY consider this transfer successful.
+     - **NOTE**: this MAY result in unrecoverable tokens if sent to an address that does not expect to receive KIP-37 tokens.
+- The above example is not exhaustive but illustrates the major points (and shows that most are shared with `safeTransferFrom` and `safeBatchTransferFrom`):
+  - Balances that are updated MUST have equivalent transfer events emitted.
+  - A receiver address has to be checked if it is a contract and if so relevant `KIP37TokenReceiver` hook function(s) have to be called on it.
+  - Balances (and events associated) that are referenced in a call to an `KIP37TokenReceiver` hook MUST be updated (and emitted) before the `KIP37TokenReceiver` hook is called.
+  - The return values of the `KIP37TokenReceiver` hook functions that are called MUST be respected if they are implemented.
+  - Only non-standard transfer functions MAY allow tokens to be sent to a recipient contract that does NOT implement the necessary `KIP37TokenReceiver` hook functions. `safeTransferFrom` and `safeBatchTransferFrom` MUST revert in that case (unless it is a hybrid standards implementation see "Backwards Compatibility").
+
+##### **_Minting-creating and burning-destroying rules:_**
+
+- A mint/create operation is essentially a specialized transfer and MUST follow these rules:
```

Comment:
Please define `mint` and `create` first.

Reply:
OK